### ISSUE TITLE
Not getting double drops when cancelling the BlockDropItemEvent

### ISSUE THREAD

**Issue by mfnalex** [mcMMO-Dev/mcMMO#4441]
Hi, I am having the following issue:

I changed mcMMO's BlockDropItemEvent listener priority to LOWEST in the source code. Double drops still work fine. Now I cancel this event on a higher priority in my own plugin, everything still works fine, EXCEPT for sugar_cane and cactus (e.g. multi-block plants).

I don't see the reason for this, as normally the double drop should have already occured in the moment I am cancelling the event.

Any idea why this happens?

Is there any way I can maybe just get the double drop chance for herbalism via the mcMMO API so I can just calculate the double drops myself?

I need this for my Drop2Inventory plugin, because many people want to automatically collect sugar cane / cactus drops while not losing the ability to get double drops.

(I already asked on the discord multiple times but noone could help yet)

Thanks for any answers :)

**Comment by nossr50** [comment]
Bonus Drops in mcMMO use a custom event (which can be cancelled to prevent them from happening)

The event is McMMOItemSpawnEvent and code related to this can be found here
https://github.com/mcMMO-Dev/mcMMO/blob/f9144e624ab9d96f052dfe18ecded6703a2f33ff/src/main/java/com/gmail/nossr50/util/Misc.java#L178-L192

A note on event listeners, changing something to LOWEST means it will be called first, the higher priorities are called last and therefor get "last say", monitor is supposed to be called very last and not meant to modify things but... well that's a long discussion to get into now.

**Comment by mfnalex** [comment]
Why did you instantly close this? This doesn't resolve the issue. When the BlockDropItemEvent is cancelled, or the drops list is cleared, the McMMOItemSpawnEvent isn't called, so there's still no chance to detect them.

**Comment by nossr50** [comment]
If you changed it to LOWEST how can you effectively cancel it? Nothing would run before it except another event on LOWEST. I explained this in my message above, I closed this because I provided a better solution to your problem than the one you were attempting (which is to intercept our custom event).

**Comment by mfnalex** [comment]
I changed the mcMMO prio to lowest. My plugin runs on highest. Even when mcMMO runs on lowest, and I cancel the event on highest, there are no double drops.

**Comment by nossr50** [comment]
@mfnalex You are misunderstanding the order of event priority, if mcMMO is listening on lowest then that means it recieves the event before your plugin does, and at that point its not cancelled, then when your plugin receives that event it cancels the event and that has no effect on mcMMO since mcMMO already processed the event.

The order goes like this

LOWEST (first in line to get an event, this is when your modified mcMMO receives the event)
LOW
NORMAL
HIGH
HIGHEST (your plugin receives the event at this point)
MONITOR

You need to have mcMMO run at a priority of at least LOW->MONITOR to effectively cancel it and have mcMMO respect that, a priority of LOWEST won't achieve this, if you put the listener priority for mcMMO to NORMAL on this event and then put your listener priority at LOWEST or LOW it would run before mcMMO and cancel it, then mcMMO would respect that.

**Comment by mfnalex** [comment]
I don't see where I am misunderstanding anything. If mcMMO runs on LOWEST, it should do the double drops. Now I cancel the event on HIGHEST, after the double drops should have already occured, but that's not the case. Am I retarded lol? ^^

**Comment by mfnalex** [comment]
I want mcMMO and to do the double drops, and AFTER the double drops have occured, I want to cancel the event

**Comment by mfnalex** [comment]
> A note on event listeners, changing something to LOWEST means it will be called first, the higher priorities are called last and therefor get "last say", monitor is supposed to be called very last and not meant to modify things but... well that's a long discussion to get into now.

btw: Dropping items into the world using World#dropItem or World#dropItemNaturally also isn't the best way to manipulate drops since the BlockDropItemEvent already has a manipulatable list of drops that's meant for exactly this thing^^

**Comment by nossr50** [comment]
> I want mcMMO to do the double drops, and AFTER the double drops have occured, I want to cancel the event

Ah, sorry for misunderstanding you.

>btw: Dropping items into the world using World#dropItem or World#dropItemNaturally also isn't the best way to manipulate drops since the BlockDropItemEvent already has a manipulatable list of drops that's meant for exactly this thing^^

Thanks for letting me know

I guess I don't understand your original question, could you rephrase it.

Specifically this question

>I changed mcMMO's BlockDropItemEvent listener priority to LOWEST in the source code. Double drops still work fine. Now I cancel this event on a higher priority in my own plugin, everything still works fine, EXCEPT for sugar_cane and cactus (e.g. multi-block plants).

>I don't see the reason for this, as normally the double drop should have already occured in the moment I am cancelling the event.

>Any idea why this happens?

**Comment by mfnalex** [comment]
Ok, let me try to rephrase it :)

I want to listen to the BlockDropItemEvent, so I can cancel it and add the item that would normally be dropped directly to the players inventory. To do so, I need to use a very high event priority so I can catch custom drops made by other plugins.

It works fine for all drops by every plugin as long as my event priority is higher (that's why I changed mcMMO's priority to LOWEST). However, multiblock plants still don't work because mcMMO does some things with multiblock plants that I don't fully understand.

If you like, we could also just talk via Discord?

**Comment by nossr50** [comment]
@mfnalex I do prefer to talk on discord in the support channels, I'm getting tired but I'll be on for a while longer. You can always catch me tomorrow.

**Comment by nossr50** [comment]
https://discord.gg/mcmmo

**Comment by mfnalex** [comment]
I tried to catch you on your discord but it didn't work out so well :D I'll get back to you tomorrow, thanks for answering :)

**Comment by nossr50** [comment]
Tag me the next time you are in there, I try to be active between 10AM-4PM Mon-Fri

**Comment by mfnalex** [comment]
Alright, see you there!

**Comment by nossr50** [comment]
I just looked into your suggestion, unfortunately the list is mutable but adding new items is not legal, from the javadocs 

>Gets list of the Item drops caused by the block break. This list is mutable - removing an item from it will cause it to not drop. It is not legal however to add new items to the list.
Returns:
The Item the block caused to drop

**Comment by mfnalex** [comment]
Wuuuut? Okay, I didn't know that. However I am still wondering why the double drops do not occur when this event is cancelled AFTER it was handled by mcMMO, I must have missed something very obvious :/

BTW you said 10AM-4PM Mon-Fri, but what timezone? :D

**Comment by nossr50** [comment]
I'm in Oregon so PDT/PST

I'm not always on, currently dealing with depression but I'm doing my best, you can always tag me and I'll reply when I can
I looked over herbalism code and it should be making use of the BlockItemDropEvent, could be a bug in your code? I'm not sure.

**Comment by mfnalex** [comment]
I wish you all the best bro!

I will try to create a very small "test" plugin to debug this in the upcoming days, if it still occurs there I'll tag you on Discord :)
